src/Adapter/MSTest.CoreAdapter/Resources/Resource.resx

```diff
@@ -368,4 +368,19 @@ Error: {1}</value>
   <data name="OlderTFMVersionFoundClassCleanup" xml:space="preserve">
     <value>An older version of MSTestV2 package is loaded in assembly, test cleanup methods might not run as expected. Please make sure all your test projects references MSTest pacakges newer then version 2.2.8.</value>
   </data>
+  <data name="CannotExpandIDataSourceAttribute" xml:space="preserve">
+    <value>Exception occurred while expanding IDataSource rows from attribute on "{0}.{1}": {2}</value>
+    <comment>{0}: TypeName with namespace, 
+{1}: Method name,</comment>
+  </data>
+  <data name="CannotExpandIDataSourceAttribute_DuplicateDisplayName" xml:space="preserve">
+    <value>Display name "{2}" on indexes {0} and {1} are duplicate.</value>
+    <comment>{0}, {1}: Zero based index if an element inside of an array
+{2}: Display name of a test.</comment>
+  </data>
+  <data name="CannotExpandIDataSourceAttribute_CannotSerialize" xml:space="preserve">
+    <value>Data on index {0} for "{1}" cannot be serialized. All data provided to thru a IDataSource should be serializable, shouldn't lose it's state when deserialized. If you need to test non-serializable data sources, please make sure you define `TestDataSourceDiscovery` attribute on your test assembly and set discovery option to `DuringExecution`.</value>
```

Comment:
```suggestion
    <value>Test "{1}" with data row index {0} cannot be serialized. All data provided through "IDataSource" should be serializable. If you need to test non-serializable data sources, please make sure you add "TestDataSourceDiscovery" attribute on your test assembly and set the discovery option to "DuringExecution".</value>
```

Reply:
It's the data which we can't serialize, not the test, and it's not always a DataRow. That was the reason I didn't use the word `row`.